specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -0,0 +1,158 @@
+`faas.trigger.name` | e.g. `mybucket` | Use `record.s3.bucket.name`
+`faas.trigger.id` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Use `record.s3.bucket.arn`
+`faas.trigger.region` | e.g. `us-east-1` | Use `record.awsRegion`
+`faas.trigger.version` | e.g. `2.1` | Use `record.eventVersion`
+`faas.execution` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Format: `${record.s3.bucket.arn}/${record.eventName}`
+
+
+## Init spans
+With both API Gateway versions the `event` objects (may) contain a timestamp of the original API Gateway request. The difference between the API Gateway timestamp and the actual start timestamp of the Lambda function is a good estimate for the initialization phase of the lambda function. 
+The agents should use this information (if available) to adapt the transaction and spans structure of the Lambda invocation to represent the initialization time:
+
+```
+...
+  [XXXXXXXXXXXXX POST api.gateway XXXXXXXXXXXXXXXXXXXXX]   // client calling the API Gateway
+     [XXXXXXXXXX MyLambdaFunction XXXXXXXXXXXXXXXXXXXX]    // Lambda Transaction
+     [XXX Lambda init XXXXX]                               // Init span
+                            [XXXX Lambda handler XXXX]      // span for the handler method
+                                    [XXXXXXXXXX]           // any Lambda subspans
+     ^                      ^
+     |                      |
+API Gateway               Timestamp of the
+RequestTimestamp          handler method start
+```
+
+To achieve the above transaction / span structure we need to backdate the begin of the created transaction using the API Gateway timestamp. We create a artificial span `Lambda init` using the API Gateway timestamp as start and the handler method timestamp as end. The `Lambda handler` span represents the actual execution of the Lambda handler method. 
+
+### Retrieving the API Gateway V1 timestamp
+With API Gateway version 1 there is no dedicated request timestamp field. 
+However, AWS adds a X-Ray tracing ID to the headers which encodes a request timestamp in **seconds**.
+Though the resolution of that timestamp is very coarse grained, in some cases it still can provide at least a rough estimate of the init phase (if the init phase is very long).
+
+The agents should use this timestamp and apply the above transaction structure if:
+- the corresponding header is available 
+- AND: `x_ray_api_gateway_request_timestamp` + 1 second < `handler_method_start_timestamp`
+
+The X-Ray timestamp can be retrieved from the HTTP header `X-Amzn-Trace-Id`. The trace id has the following format: 
+```
+1-58406520-a006649127e371903a2de979
+```
+
+The second segment (here: `58406520`) is a hexadecimal encoded timestamp in seconds.
+
+### Retrieving the API Gateway V2 timestamp
+With API Gateway version 2 the passed `event` object contains a field `timeEpoch` field (under `request context`) that denotes the request timestamp in milliseconds. 
```

Comment:
In python it's formed as

```
event["requestContext"]["requestTimeEpoch"]
```

Do those vary greatly from language to language? If it matches I think we should include the actual references (`requestContext` vs `request context`).

Reply:
since it's derived from the same JSON for all languages, I guess it's (more or less) the same naming everywhere. Will fix it in the spec.

Reply:
Per https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format it looks like it will be that for all languages.